pilot/pkg/autoregistration/controller.go

```diff
@@ -358,16 +359,27 @@ func (c *Controller) changeWorkloadEntryStateToConnected(entryName string, proxy
 	if wle == nil {
 		return false, fmt.Errorf("failed updating WorkloadEntry %s/%s: WorkloadEntry not found", proxy.Metadata.Namespace, entryName)
 	}
+
+	// check if this was actually disconnected AFTER this connTime
+	// this check can miss, but when it does the `Update` will fail due to versioning
+	// and retry. The retry includes this check and passes the next time.
+	if timestamp, ok := wle.Annotations[annotation.IoIstioDisconnectedAt.Name]; ok {
+		disconnTime, _ := time.Parse(timeFormat, timestamp)
+		if conTime.Before(disconnTime) {
+			// we slowly processed a connect and disconnected before getting to this point
+			return false, nil
+		}
+	}
+
 	lastConTime, _ := time.Parse(timeFormat, wle.Annotations[annotation.IoIstioConnectedAt.Name])
 	// the proxy has reconnected to another pilot, not belong to this one.
 	if conTime.Before(lastConTime) {
 		return false, nil
 	}
 	// Try to patch, if it fails then try to create
-	_, err := c.store.Patch(*wle, func(cfg config.Config) (config.Config, kubetypes.PatchType) {
-		setConnectMeta(&cfg, c.instanceID, conTime)
-		return cfg, kubetypes.MergePatchType
-	})
+	updated := wle.DeepCopy()
```

Comment:
nit: misleading comment about patching in L379

Reply:
Good catch